cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -714,20 +1089,26 @@ class AsofJoinNode : public ExecNode {
     DCHECK_EQ(output, outputs_[0]);
     StopProducing();
   }
-  void StopProducing() override { finished_.MarkFinished(); }
+  void StopProducing() override {
+    // avoid finishing twice, to prevent "Plan was destroyed before finishing" error
+    if (finished_.state() == FutureState::PENDING) {
+      finished_.MarkFinished();
+    }
+  }
```

Comment:
The original code just called `finished_.MarkFinished()` and I observed a failure, with the documented message as well as the message `"Future already marked finished"` in a debugging session,, that I haven't figured out. I needed some way to get runs to complete and this led me to the code-change here to avoid finishing twice. I don't know enough about the stopping logic myself to say what is a clean solution here; probably both the original code and the code-change here are not as clean as we'd like.